readthedocs/telemetry/collectors.py

```diff
@@ -0,0 +1,306 @@
+        """
+        Get the current operating system.
+
+        The output of ``lsb_release --description`` is in the form of::
+
+            Description:	Ubuntu 20.04.3 LTS
+        """
+        code, stdout, _ = self.run("lsb_release", "--description")
+        stdout = stdout.strip()
+        if code == 0 and stdout:
+            prefix = "Description:"
+            if stdout.startswith(prefix):
+                # pep8 and blank don't agree on having a space before :.
+                stdout = stdout[len(prefix) :].strip()  # noqa
+            return stdout
+        return ""
+
+    def _get_all_apt_packages(self):
+        """
+        Get all installed apt packages and their versions.
+
+        The output of ``dpkg --status`` is the form of::
+
+            Package: adduser
+            Status: install ok installed
+            Priority: important
+            Section: admin
+            Version: 3.118ubuntu2
+            Depends: passwd, debconf (>= 0.5) | debconf-2.0
+            Suggests: liblocale-gettext-perl, perl, ecryptfs-utils (>= 67-1)
+            Conffiles:
+            /etc/deluser.conf 773fb95e98a27947de4a95abb3d3f2a2
+            Description: add and remove users and groups
+            This package includes the 'adduser' and 'deluser' commands for creating
+            and removing users.
+            .
+            - 'adduser' creates new users and groups and adds existing users to
+                existing groups;
+            - 'deluser' removes users and groups and removes users from a given
+                group.
+            .
+
+            Package: apt
+            Status: install ok installed
+            Priority: important
+            Section: admin
+            Installed-Size: 4209
+            Architecture: amd64
+            Version: 2.0.6
+            Replaces: apt-transport-https (<< 1.5~alpha4~), apt-utils (<< 1.3~exp2~)
+            Provides: apt-transport-https (= 2.0.6)
+            Depends: adduser, gpgv | gpgv2 | gpgv1, libapt-pkg6.0 (>= 2.0.6)
+
+            Package: base-files
+            Essential: yes
+            Status: install ok installed
+            Priority: required
+            Section: admin
+            Installed-Size: 392
+            Architecture: amd64
+            Multi-Arch: foreign
+            Version: 11ubuntu5.5
+            Replaces: base, dpkg (<= 1.15.0), miscutils
+            Provides: base
+            Depends: libc6 (>= 2.3.4), libcrypt1 (>= 1:4.4.10-10ubuntu3)
+            Pre-Depends: awk
+            Breaks: debian-security-support (<< 2019.04.25), initscripts (<< 2.88dsf-13.3)
+
+        .. note:: Some lines have been removed for brevity.
+        """
+
+        def extract_value(key, line):
+            if line.startswith(key):
+                # pep8 and blank don't agree on having a space before :.
+                return line[len(key) :].strip()  # noqa
+            return ""
+
+        code, stdout, _ = self.run("dpkg", "--status")
```

Comment:
Didn't find an option to set the output format of dpkg, using grep and regexes is basically the same as iterating line by line, it is just hid in the regex engine. I'd also avoid doing bash to manipulate the data, using python makes it easier to test.

Reply:
I doubt that it will be fast in Python, than grepping. Also, the code is more complex when doing the manipulation in Python than with the regex. I'm not strong on this since none of them is the _correct_ solution.